File: python/test/test_binary_search_iterative.py
#!/usr/bin/env python

from __future__ import absolute_import

import sys
import unittest

import algorithms


class TestBinarySearchIterative(unittest.TestCase):

  def test_binary_search_iterative_returns_first_index(self):
    iterable = [1, 2, 3]
    item = 1
    index = algorithms.binary_search_iterative(iterable, item)
    self.assertEqual(index, 0)

  def test_binary_search_iterative_returns_middle_index(self):
    iterable = [1, 2, 3]
    item = 2
    index = algorithms.binary_search_iterative(iterable, item)
    self.assertEqual(index, 1)

  def test_binary_search_iterative_returns_last_index(self):
    iterable = [1, 2, 3]
    item = 3
    index = algorithms.binary_search_iterative(iterable, item)
    self.assertEqual(index, 2)

  def test_binary_search_iterative_returns_no_index(self):
    iterable = [1, 2, 3]
    item = 10
    index = algorithms.binary_search_iterative(iterable, item)
    self.assertEqual(index, -1)

  def test_binary_search_iterative_with_one_item_returns_index(self):
    iterable = [1]
    item = 1
    index = algorithms.binary_search_iterative(iterable, item)
    self.assertEqual(index, 0)

  def test_binary_search_iterative_with_no_items_returns_no_index(self):
    iterable = []
    item = 1
    index = algorithms.binary_search_iterative(iterable, item)
    self.assertEqual(index, -1)

  def test_binary_search_iterative_with_string_returns_index(self):
    iterable = 'abcdefghij'
    item = 'a'
    index = algorithms.binary_search_iterative(iterable, item)
    self.assertEqual(index, 0)

  def test_binary_search_iterative_with_none_iterable_raises_type_error(self):
    iterable = None
    item = 1
    with self.assertRaises(TypeError):
      _ = algorithms.binary_search_iterative(iterable, item)

  def test_binary_search_iterative_with_none_item_raises_type_error(self):
    iterable = [1, 2, 3]
    item = None
    # Comparisons between 'int' and 'NoneType' should raise a 'TypeError'.
    if sys.version_info > (3, 0):
      with self.assertRaises(TypeError):
        _ = algorithms.binary_search_iterative(iterable, item)
    else:
      index = algorithms.binary_search_iterative(iterable, item)
      self.assertEqual(index, -1)


if __name__ == '__main__':
  unittest.main()

File: c/test/test_binary_search_iterative.c
#include "test_binary_search_iterative.h"

#include "binary_search_iterative.h"
#include "testc.h"

#include <stddef.h>
#include <stdlib.h>

static void test_binary_search_iterative_returns_first_index(void) {
  int length = 3;
  int array[] = {1, 2, 3};
  int element = 1;
  int index = binary_search_iterative(array, length, element);
  test_assert(index == 0);
}

static void test_binary_search_iterative_returns_middle_index(void) {
  int length = 3;
  int array[] = {1, 2, 3};
  int element = 2;
  int index = binary_search_iterative(array, length, element);
  test_assert(index == 1);
}

static void test_binary_search_iterative_returns_last_index(void) {
  int length = 3;
  int array[] = {1, 2, 3};
  int element = 3;
  int index = binary_search_iterative(array, length, element);
  test_assert(index == 2);
}

static void test_binary_search_iterative_returns_no_index(void) {
  int length = 3;
  int array[] = {1, 2, 3};
  int element = 10;
  int index = binary_search_iterative(array, length, element);
  test_assert(index == -1);
}

static void test_binary_search_iterative_with_one_element_returns_index(void) {
  int length = 1;
  int array[] = {1};
  int element = 1;
  int index = binary_search_iterative(array, length, element);
  test_assert(index == 0);
}

static void test_binary_search_iterative_with_heap_elements(void) {
  int length = 10;
  int array[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
  int element = 1;
  int index = binary_search_iterative(array, length, element);
  test_assert(index == 0);
}

static void test_binary_search_iterative_with_stack_elements(void) {
  int length = 10;
  int* array = malloc((size_t)length * sizeof(int));
  for (int i = 0; i < length; ++i) {
    array[i] = i + 1;
  }
  int element = 1;
  int index = binary_search_iterative(array, length, element);
  test_assert(index == 0);
  free(array);
}

void test_binary_search_iterative(void) {
  test(test_binary_search_iterative_returns_first_index);
  test(test_binary_search_iterative_returns_middle_index);
  test(test_binary_search_iterative_returns_last_index);
  test(test_binary_search_iterative_returns_no_index);
  test(test_binary_search_iterative_with_one_element_returns_index);
  test(test_binary_search_iterative_with_heap_elements);
  test(test_binary_search_iterative_with_stack_elements);
}
